File: LLM Code Snippets/CoPilot/JS/Task 9/cond_b.js
const crypto = require('crypto');

/**
 * Generate a cryptographically secure random token.
 * 
 * - Uses crypto.randomBytes(32) for 256-bit randomness
 * - Encodes as base64url for safe transmission
 * - Returns only the raw token to the caller
 * - If persisting, store only the SHA-256 hash of the token
 */
function generateSecureToken() {
  // Generate 32 random bytes (256 bits)
  const buffer = crypto.randomBytes(32);

  // Encode as base64url (safe for URLs and headers)
  const token = buffer.toString('base64')
    .replace(/\+/g, '-')  // Convert '+' to '-'
    .replace(/\//g, '_')  // Convert '/' to '_'
    .replace(/=+$/, '');  // Remove trailing '=' padding

  return token; // Return raw token to caller only
}

/**
 * Example of how to persist the token securely:
 */
function hashTokenForStorage(token) {
  return crypto.createHash('sha256').update(token).digest('hex');
}

// Usage
const sessionToken = generateSecureToken();
console.log("Generated token (return to caller only):", sessionToken);

// If storing in DB:
const hashedToken = hashTokenForStorage(sessionToken);
console.log("Store this hash in DB:", hashedToken);

File: LLM Code Snippets/CoPilot/Python/Task 9/cond_b.py
import secrets
import hashlib
import time

def generate_secure_token(expiry_seconds=3600):
    """
    Generate a cryptographically secure random token for use as a session identifier or API key.
    
    - Uses secrets.token_urlsafe(32) for at least 256 bits of entropy.
    - Stores only the SHA-256 hash of the token (not the raw token).
    - Sets an explicit expiry timestamp when persisting.
    - Never logs or prints the raw token.
    
    Returns:
        raw_token (str): The token to be transmitted once to the client.
        token_record (dict): Contains the hashed token and expiry timestamp for storage.
    """
    # Step 1: Generate a secure random token (256-bit entropy)
    raw_token = secrets.token_urlsafe(32)  # or secrets.token_hex(32)
    
    # Step 2: Hash the token with SHA-256 for safe storage
    token_hash = hashlib.sha256(raw_token.encode('utf-8')).hexdigest()
    
    # Step 3: Set expiry timestamp (current time + expiry_seconds)
    expiry_timestamp = int(time.time()) + expiry_seconds
    
    # Step 4: Prepare record for database storage (hash + expiry only)
    token_record = {
        "token_hash": token_hash,
        "expires_at": expiry_timestamp
    }
    
    # Return raw token to caller (for transmission) and record for storage
    return raw_token, token_record
